### PR TITLE
Introduce Helmfile's own filesystem abstraction to correctly unit test some components

### PR DESCRIPTION
**Pull request by arkaitzj**
This is a successor to #304
Follow-up for #297

### CODE REVIEW

**Comment by yxxhero** [comment]
@arkaitzj @mumoshu I think this abstraction is very good. It can simplify the code and I look forward to completing it.

**Comment by arkaitzj** [comment]
@mumoshu, @yxxhero what do you think? it passes all tests and works the same as main branch on my convoluted use case.

File: pkg/filesystem/fs.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	FileExistsAt      func(string) bool
+	DirectoryExistsAt func(string) bool
+	Stat              func(string) (os.FileInfo, error)
+	Getwd             func() (string, error)
+	Chdir             func(string) error
+	Abs               func(string) (string, error)
+}
+
+func DefaultFileSystem() *FileSystem {
+	fs := FileSystem{
```
Comment (yxxhero) [inline pkg/filesystem/fs.go:24]: `fs` var has the same name for `io/fs`. Can you change it. Thanks very much.

File: pkg/filesystem/fs.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+		dfs.Chdir = params.Chdir
+	}
+	if params.Abs != nil {
+		dfs.Abs = params.Abs
+	}
+
+	return dfs
+}
+
+func (filesystem *FileSystem) fileExistsAtDefault(path string) bool {
```
Comment (yxxhero) [inline pkg/filesystem/fs.go:81]: add some test for `fs.go`.

**Comment by yxxhero** [comment]
@arkaitzj 
1. sqush your commmits
2. add more test for `fs.go`
Thanks very much.

**Comment by yxxhero** [comment]
@arkaitzj @mumoshu  On the whole, the core logic has not changed, but the relevant operations of the file system are abstracted into a separate package, which simplifies the parameter transfer of many functions. I think it would be better to add some unit tests.

**Comment by yxxhero** [comment]
@arkaitzj @mumoshu 
1. Overall, it looks good to me
2. same as main branch on convoluted use case for @arkaitzj

**Comment by mumoshu** [comment]
@arkaitzj Whoa, well done! This is absolutely nice. I especially love your decision to bring our own fs interface, rather than building on top of Golang's native fs interfaces. It turned out to be simpler and requires less changes for our use-case.

I've rebased your work on top of our current state of the main branch.
I'll see if CI passes at this point.

In the meantime, could you let me confirm- is this still WIP? If not, just let us know so that we can fully review this! Thanks.

**Comment by arkaitzj** [comment]
No longer wip.
So far I am looking to get it merged. 
Thanks!

**Review by mumoshu — APPROVED** [review]
LGTM. Thank you so much for your efforts and contribution @arkaitzj!!

**Pull request merged**